### PR TITLE
Adding simulated benchmark stuff.

### PR DESCRIPTION
**Pull request by Yash9060**
Internal issue b/377888771

This cl adds code used to benchmark, fsspec vs dataflux (simulated version) 

Adding readme and benchmarks result.

- [X] Tests pass
- [X] Appropriate changes to documentation are included in the PR

### CODE REVIEW

File: dataflux_pytorch/benchmark/checkpointing/simulated/benchmark-deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+              schedulerName: default-scheduler
+              restartPolicy: Never
+
+              priorityClassName: medium
+              hostNetwork: true
+              dnsPolicy: ClusterFirstWithHostNet
+              terminationGracePeriodSeconds: 30
+              containers:
+              - name: {CONTAINER_NAME}
+                image: gcr.io/{PROJECT_NAME/{CONTAINER_NAME}
```
Comment (jdnurme) [inline dataflux_pytorch/benchmark/checkpointing/simulated/benchmark-deployment.yaml:45]: Missing closing bracket on PROJECT_NAME

File: dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    """Initializes the distributed environment."""
+    world_size = int(os.environ["WORLD_SIZE"])
+    job_index = int(os.environ.get("JOB_INDEX", 0))
+    job_completion_index = int(os.environ.get("JOB_COMPLETION_INDEX", 0))
+    processes_in_job = int(os.environ.get("PROCESSES_IN_JOB", 1))
+    rank = job_index * processes_in_job + job_completion_index
+    os.environ["NODE_RANK"] = str(rank)
+
+    configure_master_addr()
+    torch.distributed.init_process_group(backend='gloo',
```
Comment (jdnurme) [inline dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py:47]: Maybe a brief comment here about gloo backend specifying cpu usage.

**Review by jdnurme — APPROVED** [review]

File: dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py (hunk 2 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py:48]: Done

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+*   This is a simulated version of the code that runs on CPU to test the performance of GCS as a backing store for AI/ML workloads without GPU.
+
+*   The script generates [state_dict](https://pytorch.org/tutorials/recipes/recipes/what_is_state_dict.html#what-is-a-state-dict-in-pytorch) that represents the state of the model. Each node then saves a part of the state_dict to GCS (the state_dict is divided among the nodes, such that the size of each shard is the same). During restore, each node reads its shard from GCS in parallel. The `.metadata` file, which is written and read by the Rank_0 (coordinator) node, contains information about which shard each node should read.
+
+*   The official implementation of FSDP in PyTorch Lightning for save_checkpoint can be found [here]
+(https://github.com/Lightning-AI/pytorch-lightning/blob/3627c5bfac704d44c0d055a2cdf6f3f9e3f9e8c1/src/lightning/fabric/strategies/fsdp.py#L419)
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/README.md:10]: This link doesn't show up correctly on the preview I think because it's split across lines

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+*   The official implementation of FSDP in PyTorch Lightning for load_checkpoint can be found [here](https://github.com/Lightning-AI/pytorch-lightning/blob/3627c5bfac704d44c0d055a2cdf6f3f9e3f9e8c1/src/lightning/fabric/strategies/fsdp.py#L519)
+
+*   While the official implementation of both `save_checkpoint` and `load_checkpoint` does many things beyond just saving/loading the state_dict, these additional steps have been omitted for the following reasons: 
+*   *   We wanted to demonstrate how GCS will scale with increasing model/checkpoint sizes without needing an actual GPU.
+*   *   These additional steps do not interact with GCS.
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/README.md:15]: These lines show up with 2 bullet points on preview

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+```shell
+gcloud config set project {PROJECT_ID}
+```
+
+### Environment variables:
+
+Set the following environment variables by updating the deployment file to run on a GKE cluster.
+1. Set the environment variables required to run the demo. These include:
+  
+  * `PROJECT`: The GCP project you're using.
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/README.md:34]: nit: I'm not sure the indentation of this list is showing up correctly on preview. You may need 4 spaces instead of 2

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  
+  * `CKPT_DIR_PATH`: The full path of the directory in which to save checkpoints, in the format `gs://<bucket>/<directory>/`
+
+  * `WORLD_SIZE`: The number of nodes.
+
+  * `LAYER_SIZE`: The size of each layer.
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/README.md:40]: Is this really accurate that it's the size? I thought it was actually the count of layers, in which case the naming may need to be updated. In either case, can you add some more information about what this means?

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+2. Set the optional environment variables, if desired:
+  
+  * `PADDING_SIZE`: The number of dummy tensors to add to the state_dict in order to produce checkpoints of desired size.
+    *   Both padding_size and layer_size will impact the size of the checkpoint. Therefore, set layer_size to an appropriate value first, and then adjust padding_size until the checkpoint of the desired size is generated. The default value for padding_size is 4000.
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/README.md:45]: Could you add some more info about how LAYER_SIZE and PADDING_SIZE are related? For example if I double LAYER_SIZE and 10x PADDING_SIZE, how much bigger is the resulting checkpoint?

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+*   If a user wishes to run an end to end FSDP implementation on an actual GPU, then please look at the example under the [multinode](https://github.com/GoogleCloudPlatform/dataflux-pytorch/tree/main/demo/lightning/checkpoint/multinode) directory.
+
+## Getting started
+
+### Configuration
+
+Ensure you are running within a python virtual environment, and ensure gcloud config project is set to your project. You can use the following command to set it:
+
+```shell
+gcloud config set project {PROJECT_ID}
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/README.md:26]: This command is also given below

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+docker tag {YOUR_CONTAINER_NAME} gcr.io/{YOUR-GCP-PROJECT}/{YOUR_CONTAINER_NAME}
+docker push gcr.io/{YOUR-GCP-PROJECT}/{YOUR_CONTAINER_NAME}
+```
+
+4.  Update the values in `benchmark-deployment.yaml`.
+
+5.  Deploy the configuration using the following command:
+
+```shell
+kubectl apply -f /app/dataflux_pytorch/benchmark/checkpointing/simulated/benchmark-deploy.yaml
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/README.md:97]: If the user's working directory is `dataflux-pytorch` I don't think this path will have a leading `/app/`

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+
+### Benchmarking Results.
+
+The table below contains benchmarking times for saving checkpoints to GCS. The average save/load time is taken over 8 calls to  save_checkpoint and load_checkpoint. The tests were conducted on a single GKE cluster containing several nodes, with each VM having the `n2-standard-32` configuration. The GKE cluster, and the GCS bucket were located in `us-central1`.
+
+### Checkpoint Save & Load times.
+
+
+| Num_nodes | Parameters | Checkpoint Type | Num_layers + padding_size | Size of each shard (in GB) | Total size (in GB) | Save time (in seconds) | Load time (in seconds) | Average Throughput (in GBps) |
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/README.md:108]: The average throughput column doesn't really make sense here because it's not clear if it's referring to save or load. You may want to separate this out to 2 different tables for save/load like we've done in some of the other docs.

File: dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py:1]: multi_node_benchmark.py is already here, why is this added?

**Review by MattIrv — CHANGES_REQUESTED** [review]

**Review by awonak — COMMENTED** [review]
Nice work and great documentation! Feel free to ignore any of my opinionated markdown formatting suggestions.

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 8 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:26]: Removed it from Auth section.

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 10 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:108]: Oh I forgot the egress column, I think splitting it into 2 will be better fir readability. 

Done

File: dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py:1]: Wanted to rename multi_node_benchmark.py to just benchmark.py. Removed multi_node_benchmark.py

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+Run the following commands to authenticate your application so it can access GCS buckets.
+```shell
+gcloud config set project PROJECT_ID
+gcloud auth application-default login
+```
+
+### Running
+
+_Note: The following instructions assume that you have Jobset and Kueue enabled on your GKE cluster. For easier compatibility, we recommend creating a cluster using [XPK](https://github.com/google/xpk) which will configure these features automatically._
```
[the remark this answers is not in the document]
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:70]: Done

File: dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+def run_benchmark(world_size: int, layer_size: int, project: str,
+                  filepath: str, padding_size: int, sample_count: int,
+                  use_fsspec: bool) -> None:
+    rank = init_processes() if os.environ.get("COORDINATOR_ADDRESS") else 0
+
+    benchmark_strategy = BenchmarkStrategy(project=project,
+                                           path=filepath,
+                                           use_fsspec=use_fsspec)
+    # According to `create_default_local_load_plan` https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/default_planner.py#L343
```
[the remark this answers is not in the document]
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py:62]: Great suggestion.

File: dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    processes_in_job = int(os.environ.get("PROCESSES_IN_JOB", 1))
+    rank = job_index * processes_in_job + job_completion_index
+    os.environ["NODE_RANK"] = str(rank)
+
+    configure_master_addr()
+    # Using gloo backend since the simulated version runs on CPU.
+    torch.distributed.init_process_group(backend='gloo',
+                                         rank=rank,
+                                         world_size=world_size)
+    return rank
```
[the remark this answers is not in the document]
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py:51]: Changed it to not return rank.

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+```
+### GCS Auth
+
+Run the following commands to authenticate your application so it can access GCS buckets.
+```shell
+gcloud config set project PROJECT_ID
+gcloud auth application-default login
+```
+
+### Running
```
[the remark this answers is not in the document]
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:68]: Done

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+  * `USE_FSSPEC`: If set to true, the code will use [gcsfs/fsspec](https://github.com/fsspec/gcsfs) in order to save_checkpoint/restore_checkpoint from GCS.
+    *   If not set, it will use GCS Connector for Pytorch by default.
+
+### Installing Requirements:
+ 
+* Install the required dependencies using the following commands:
+
+```shell
+ pip install -r dataflux_pytorch/benchmark/requirements.txt;
```
[the remark this answers is not in the document]
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:57]: Renoved.

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:10]: Done

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 9 above)
Comment (jdnurme) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:96]: The apply command typically isn't run by docker, but instead run locally by the user to apply a configuration to a cluster. I don't think most users would have the /app directory.

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+2. Set the optional environment variables, if desired:
+  
+  * `PADDING_SIZE`: The number of dummy tensors to add to the state_dict in order to produce checkpoints of desired size.
+    *   Both padding_size and layer_size will impact the size of the checkpoint. Therefore, set layer_size to an appropriate value first, and then adjust padding_size until the checkpoint of the desired size is generated. The default value for padding_size is 4000.
+  
+  * `SAMPLE_COUNT`: The number of times save_checkpoint/load_checkpoint is called. The default is 8.
+
+  * `USE_FSSPEC`: If set to true, the code will use [gcsfs/fsspec](https://github.com/fsspec/gcsfs) in order to save_checkpoint/restore_checkpoint from GCS.
+    *   If not set, it will use GCS Connector for Pytorch by default.
```
[the remark this answers is not in the document]
Comment (jdnurme) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md]: Thanks for including the screenshots with these changes! very helpful!

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 6 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md]: Yes it is accurate. This is layer_size. If we look at the code,  layer_size is number of neurons in each layer.  padding_size on the other hand decides the total number of layers. (But the name is still kept to padding_size since we are sort of padding it to the state_dict. 

Added some documentation around layer_size.

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 7 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md]: Done.

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 3 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md]: Done

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 9 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:98]: I just ran the the GKE demo using deployment-yaml file right now. Seems like it is working as expected (and the path is correct).

**Review by awonak — APPROVED** [review]
LGTM

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 9 above)
Comment (jdnurme) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md:98]: Typically the apply command is run from the users personal machine, is it not? What did your execution command look like? Usually the `/app` directory is just built into the image itself and is used for code that's running within the docker container.

File: dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py [hunk 17]
```diff
@@ -45,21 +44,25 @@ def init_processes() -> int:
     os.environ["NODE_RANK"] = str(rank)
 
     configure_master_addr()
+    # Using gloo backend since the simulated version runs on CPU.
     torch.distributed.init_process_group(backend='gloo',
                                          rank=rank,
                                          world_size=world_size)
     return rank
```
Comment (MattIrv) [inline dataflux_pytorch/benchmark/checkpointing/simulated/benchmark.py:51]: This still returns even though the return type is None

File: dataflux_pytorch/benchmark/checkpointing/simulated/README.md (hunk 9 above)
Comment (Yash9060) [reply dataflux_pytorch/benchmark/checkpointing/simulated/README.md]: Oh I was completely misreading the readme. Corrected. 

Thanks for catching this @MattIrv

**Review by MattIrv — APPROVED** [review]
